---
title: "CytAssist Pilot"
date: "2025-03-05"
output:
  html_document: default
params:
  spaceranger_dir: null
  coda_ann_path: null
  output_dir: "analysis_pilot"
  sample_name: "SMP"
---

# VISIUM DATA ANALYSIS WITH SEURAT
# Incorporating CODA for cell type annotations

##### Loading packages
```{r, echo=FALSE}
library(Seurat)
library(ggplot2)
library(leiden)
library(patchwork)
library(dplyr)
library(rhdf5)
library(ggrepel)
library(Rfast2)
library(EnhancedVolcano)
library(fgsea)
library(msigdbr)
library(RColorBrewer)
library(ggpubr)
library(MAST)
saveRDS(sessionInfo(), "sessionInfo.rds")
```

## Increasing the allowed memory to avoid errors regarding lack of memory
```{r}
options(future.globals.maxSize = 1024 * 1024 * 1024)  # 1 GB
```

## Loading the data for sample SMP
```{r}
SMP <- Load10X_Spatial(params$spaceranger_dir)
dim(SMP)
head(SMP@meta.data)
```

## INTEGRATING CODA CELL CLASSIFICATION INTO METADATA
```{r}
# extract SMP metadata
SMPmetadata <- SMP@meta.data

# uploading CODA annotations
codaSMP <- as.data.frame(read.csv(params$coda_ann_path))
rownames(codaSMP) <- codaSMP$X
codaSMP <- codaSMP[, -c(1, 2, 3, 4, 5, 6)]
head(codaSMP)

# merging CODA annotations to metadata
SMPmetadata <- merge(SMPmetadata, codaSMP, by = "row.names")

# assigning cell types from CODA quantifications per spot (barcode)
SMPmetadata$CODA <- "mixed"
SMPmetadata$CODA[SMPmetadata$RespondingPDAC > 51] <- "responding"
SMPmetadata$CODA[SMPmetadata$NonrespondingPDAC > 51] <- "non-responding"
SMPmetadata$CODA[SMPmetadata$Stroma > 51] <- "stroma"
SMPmetadata$CODA[SMPmetadata$Immune > 51] <- "immune"
SMPmetadata$CODA[SMPmetadata$Vasculature > 51] <- "vasculature"
SMPmetadata$CODA[SMPmetadata$Fat > 51] <- "fat"
SMPmetadata$CODA[SMPmetadata$Acini > 51] <- "acini"
SMPmetadata$CODA[SMPmetadata$Islets > 51] <- "islets"
SMPmetadata$CODA[SMPmetadata$Nerve > 51] <- "nerve"
SMPmetadata$CODA[SMPmetadata$Duct > 51] <- "normal duct"
SMPmetadata$CODA[SMPmetadata$RespondingPDAC == 0 & 
                 SMPmetadata$NonrespondingPDAC == 0 & 
                 SMPmetadata$Stroma == 0 & 
                 SMPmetadata$Immune == 0 &
                 SMPmetadata$Vasculature == 0 &
                 SMPmetadata$Fat == 0 &
                 SMPmetadata$Acini == 0 &
                 SMPmetadata$Islets == 0 &
                 SMPmetadata$Nerve == 0 &
                 SMPmetadata$Duct == 0] <- "NA"
table(SMPmetadata$CODA)

SMP <- AddMetaData(object = SMP, metadata = SMPmetadata)
head(SMP)
```

##### saving object with CODA classification
```{r}
save(SMP, file = "SMP_CODA.Rda")
```

# DATA PREPROCESSING

## QC and filtering
```{r}
load("SMP_CODA.Rda")

# proportion of mitochondrial reads
SMP[["percent_mt"]] <- PercentageFeatureSet(SMP, pattern = "^MT-")
head(SMP@meta.data)

VlnPlot(SMP, features = c("nFeature_Spatial", "nCount_Spatial", "percent_mt"), ncol = 3)
ggsave("SMP_QC1.png")

plot1 <- FeatureScatter(SMP, feature1 = "nCount_Spatial", feature2 = "percent_mt")
plot2 <- FeatureScatter(SMP, feature1 = "nCount_Spatial", feature2 = "nFeature_Spatial")
wrap_plots(plot1, plot2)
ggsave("SMP_QC2.png")

# plots to verify the variance in counts across spots
plot3 <- VlnPlot(SMP, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
plot4 <- SpatialFeaturePlot(SMP, features = "nCount_Spatial", pt.size.factor = 7) + theme(legend.position = "right")
wrap_plots(plot3, plot4)
ggsave("SMP_QC3.png")

# NOTE: data looks good quality and spots were not filtered out of the analysis
```

##### saving filtered dataset
```{r}
save(SMP, file = "SMP_filtered.Rda")
```

## NORMALIZATION WITH SCTransform
```{r}
load("SMP_filtered.Rda")

SMP <- SCTransform(SMP, assay = "Spatial", vars.to.regress = "percent_mt",verbose = F)
```

##### saving SCTransformed data
```{r}
save(SMP, file = "SMP_SCTransformed.Rda")
```

## DIMENSIONALITY REDUCTION, CLUSTERING, VISUALIZATION
```{r}
load("SMP_SCTransformed.Rda")

SMP <- RunPCA(SMP, assay = "SCT")
DimPlot(SMP, reduction = "pca")

# cluster spots
SMP <- FindNeighbors(SMP, reduction = "pca", dims = 1:30)
SMP <- FindClusters(SMP, algorithm = 4)

# run UMAP for dimensional reduction
SMP <- RunUMAP(SMP, reduction = "pca", dims = 1:30)

# creating a palette of colors that fits 15 clusters
colorcount <- length(unique(SMP$SCT_snn_res.0.8))
myPalette <- colorRampPalette(brewer.pal(12, "Paired"))

DimPlot(SMP, reduction = "umap", label = TRUE)
ggsave("SMP_Dimplot.png")
SpatialDimPlot(SMP, label = T, label.size = 2.5, pt.size.factor = 7) + 
  scale_fill_manual(values = myPalette(colorcount))
ggsave("SMP_clustering_res08.png")

# highligthing specific clusters (cluster by cluster plot)
# all clusters
SpatialDimPlot(SMP, 
               cells.highlight = CellsByIdentities(object = SMP,idents = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 
                                                                             11, 12, 13, 14, 15, 16, 17)),
               facet.highlight = T, ncol = 6, pt.size.factor = 7, image.alpha = 1, alpha = 0.5)

ggsave("SMP_cluster_highlights_res08.png")
```

##### saving reduced data
```{r}
save(SMP, file = "SMP_clustered.Rda")
```

## FINDING CLUSTER MARKER GENES
```{r}
load("SMP_clustered.Rda")

### find all markers in each cluster distinguishing one cluster from all other spots
markers_SMP <- FindAllMarkers(SMP, test.use = "MAST",only.pos = T, min.pct = 0.25, logfc.threshold = 0.25)
markers_df <- as.data.frame(markers_SMP %>% group_by(cluster) %>% slice_max(n = 10, order_by = avg_log2FC))
DoHeatmap(SMP,features = markers_df$gene)
ggsave("SMP_markersHM.png")
```

## INDENTIFYING HIGHLY VARIABLE SPATIAL FEATURES
```{r}
# FindSpatiallyVariables search for features without pre-annotation
# "markvariogram" finds genes whose expression depends on spatial location
SMP <- FindVariableFeatures(SMP, selection.method = "vst", nfeatures = 2000)
topSMP_varfeat <- head(VariableFeatures(SMP), 10)
topSMP_varfeat

SMP <- FindSpatiallyVariableFeatures(SMP, assay = "SCT", 
                                     features = VariableFeatures(SMP)[1:2000], 
                                     selection.method = "moransi")
top_features <- head(VariableFeatures(SMP, method = "moransi"), 20)
SpatialFeaturePlot(SMP, features = top_features, ncol = 5, alpha = c(0.1,1))
ggsave("SMP_varfeatures.png")
```

##### saving variable features data
```{r}
save(SMP, file = "SMP_VarFeatures.Rda")
```

## SETTING CODA AS ACTIVE IDENTITIES
```{r}
load("SMP_VarFeatures.Rda")

# making CODA annotations the active identities
Idents(SMP) <- "CODA"
table(Idents(SMP))

SpatialDimPlot(SMP, label = T, label.size = 2.5, pt.size.factor = 5, cols = "Paired")
ggsave("SMP_clustering_CODA.png")

SpatialDimPlot(SMP, 
               cells.highlight = CellsByIdentities(object = SMP,idents = c("responding", "non-responding", "stroma", 
                                                                             "immune", "vasculature",  "fat", "acini", 
                                                                             "islets", "nerve", "normal duct", "mixed")),
               facet.highlight = T, ncol = 6, pt.size.factor = 7, image.alpha = 1, alpha = 0.5)

ggsave("SMP_cluster_highlights_CODA.png")

markers_SMP <- FindAllMarkers(SMP, test.use = "MAST",only.pos = T, min.pct = 0.25, logfc.threshold = 0.25)
codamarkers_df <- as.data.frame(markers_SMP %>% group_by(cluster) %>% slice_max(n = 10, order_by = avg_log2FC))
DoHeatmap(SMP,features = codamarkers_df$gene)
ggsave("SMP_markersHM_CODA.png")
```

##### saving the object using the CODA annotations as active idents
```{r}
save(SMP, file = "SMP_CODAidents.Rda")
```

## Differential expression analysis with CODA annotations
```{r}
load("SMP_CODAidents.Rda")

DGEA <- FindMarkers(SMP, test.use = "MAST", ident.1 = "responding", ident.2 = "non-responding", logfc.treshold = 0.1, min.pct = 0.1)
dgea_df <- as.data.frame(DGEA)

top_genes <- rownames(DGEA)[order(DGEA$p_val_adj)][1:10]

EnhancedVolcano(dgea_df, 
                lab = rownames(dgea_df), 
                x = 'avg_log2FC', 
                y = 'p_val',
                title = 'responding vs. non-responding',
                pCutoff = 1e-2,
                FCcutoff = 1,
                pointSize = 0.5,
                xlim = c(-1.5, 1.5), 
                ylim = c(0, 5), 
                labSize = 4.0)
ggsave("SMP_DGEA_volcano.png")
write.csv(dgea_df, file = "SMP_DGEA_output.csv")

# GSEA
# ranking genes
ranks <- dgea_df$avg_log2FC
names(ranks) <- rownames(dgea_df)
head(ranks)
ranks <- sort(ranks, decreasing = TRUE) # sorting genes by ranking
plot(ranks)

# loading MSigDB Hallmark pathways for mouse
hs_df = msigdbr(species = "Homo sapiens", category = "H")

# converting to list format
hs_list = hs_df %>% split(x = .$gene_symbol, f = .$gs_name)

# Show the first few pathways, and within those, show only the first few genes. 
hs_list %>% 
  head() %>% 
  lapply(head)

# running fgsea
fgseaRes <- fgseaMultilevel(pathways = hs_list, 
                            stats = ranks, 
                            scoreType = "std", 
                            nPermSimple = 10000, 
                            nproc = 1)
head(fgseaRes)
save(fgseaRes, file = "SMP_fgsea_hallmarks.Rda")

ggplot(fgseaRes, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=pval<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()
ggsave("SMP_GSEA_NES_waterfall.png")
```


## Identifying PDAC transcriptional subtypes
```{r}
### AddModuleScore for PDAC subtypes (Moffitt classification - classical and basal-like)

load("SMP_CODAidents.Rda")

# subsetting for epithelial cells
SMPepit <- subset(SMP, idents = c("responding", "non-responding", "normal duct"))
SpatialDimPlot(SMPepit, label = T, label.size = 2.5, pt.size.factor = 5, cols = "Paired")

save(SMPepit, file = "SMP_Epithelial.Rda")

# PDAC basal subtype
basal_genes <- list(c('VGLL','UCA1','S100A2','LY6D','SPRR3','SPRR1B','LEMD1','KRT15',
                      'CTSL2','DHRS9','AREG','CST6','SERPINB3','KRT6C','KRT6A','SERPINB4',
                      'FAM83A','SCEL','FGFBP1','KRT7','KRT17','GPR87','TNS4','SLC2A1',
                      'ANXA8L2'))
SMPepit <- AddModuleScore(SMPepit, features = basal_genes, name = 'basal')
SpatialFeaturePlot(SMPepit, features = 'basal1', images = 'slice1', pt.size.factor = 5)
ggsave("SMP_basal.png")

VlnPlot(SMPepit, features = 'basal1')
ggsave("SMP_basal_vln.png")

# PDAC classical subtype
classical_genes <- list(c('BTNL8','FAM3D','ATAD4','AGR3','CTSE','LOC400573','LYZ','TFF2',
                          'TFF1','ANXA10','LGALS4','PLA2G10','CEACAM6','VSIG2','TSPAN8',
                          'ST6GALNAC1','AGR2','TFF3','CYP3A7','MYO1A','CLRN3','KRT20',
                          'CDH17','SPINK4','REG4'))
SMPepit <- AddModuleScore(SMPepit, features = classical_genes, name = 'classical')
SpatialFeaturePlot(SMPepit, features = 'classical1', images = 'slice1', pt.size.factor = 5)
ggsave("SMP_classical.png")

VlnPlot(SMPepit, features = 'classical1')
ggsave("SMP_classical_vln.png")
```

## Identifying CAF subtypes
```{r}
### AddModuleScore for CAF subtypes

load("SMP_CODAidents.Rda")

# subsetting for stroma
SMPstroma <- subset(SMP, idents = "stroma")
save(SMPstroma, file = "SMP_Stroma.Rda")

# panCAFs
pancaf_genes <- list(c('LUM','DCN','COL1A1','VIM','CD39','FSP1','FAP','ACTA2','PDPN'))
SMPstroma <- AddModuleScore(SMPstroma, features = pancaf_genes, name = 'panCAF')
SpatialFeaturePlot(SMPstroma, features = 'panCAF1', images = 'slice1', pt.size.factor = 5)
ggsave("SMP_panCAF.png")

# iCAFs
iCAF_genes <- list(c('CXCL1','CXCL2','CCL2','CXCL12','PDGFRA','CFD','LMNA','DPT','HAS1','HAS2'))
SMPstroma <- AddModuleScore(SMPstroma, features = iCAF_genes, name = 'iCAF')
SpatialFeaturePlot(SMPstroma, features = 'iCAF1', images = 'slice1', pt.size.factor = 5)
ggsave("SMP_iCAF.png")

# myCAFs
myCAF_genes <- list(c('TAGLN','MYL9','TPM2','MMP11','POSTN','HOPX','TWIST1','SOX4'))
SMPstroma <- AddModuleScore(SMPstroma, features = myCAF_genes, name = 'myCAF')
SpatialFeaturePlot(SMPstroma, features = 'myCAF1', images = 'slice1', pt.size.factor = 5)
ggsave("SMP_myCAF.png")

# apCAFs
apCAF_genes <- list(c('HLA-DRA','HLA-DPA1','CD74','HLA-DQ','SLPI'))
SMPstroma <- AddModuleScore(SMPstroma, features = apCAF_genes, name = 'apCAF')
SpatialFeaturePlot(SMPstroma, features = 'apCAF1', images = 'slice1', pt.size.factor = 5)
ggsave("SMP_apCAF.png")
```

### Module score analysis for Hallmark Pathways for Epithelial (responding, non-responding, normal) Cells
```{r}
load("SMP_Epithelial.Rda")

# loading and create a list of MSigDB Hallmark pathways
hallmark_pathways = msigdbr(species = "Homo sapiens", category = "H")
hallmark_list <- split(hallmark_pathways$gene_symbol, hallmark_pathways$gs_name)


# loop for calculating module scores
for (pathway in names(hallmark_list)) {
  SMPepit <- AddModuleScore(
    object = SMPepit,
    features = list(hallmark_list[[pathway]]),
    name = pathway,
    ctrl = 5
  )
  
  # Plot the SpatialFeaturePlot
    plot <- SpatialFeaturePlot(SMPepit, features = paste0(pathway, "1"))
    print(plot)
  
  # Save the plot
    ggsave(filename = file.path(paste0(pathway, "_feature_plot.png")), plot = plot, width = 8, height = 6)
}

head(SMPepit@meta.data)

save(SMPepit, file = 'SMPepit_modscores.Rda')
```

### module scores violin plots
```{r}
# subsetting data for violin plots
epit_meta <- SMPepit@meta.data

# creating groups for statistical comparisons
mycomparisons <- list(c('normal duct','responding'), c('normal duct','non-responding'), c('responding','non-responding'))

# loop for module scores
for (pathway in names(hallmark_list)) {
  pathway_name <- paste0(pathway, "1")
  
  # Check if the pathway score exists in the metadata
  if (pathway_name %in% colnames(epit_meta)) {
    p <- ggplot(epit_meta, aes(x = CODA, y = .data[[pathway_name]])) + 
      geom_violin(aes(fill = CODA)) +
      geom_jitter(shape = 16, position = position_jitter(0.2)) +
      scale_x_discrete(name = 'Cell Type', limits = c("normal duct", "responding", "non-responding")) + 
      stat_compare_means(comparisons = mycomparisons) +
      ggtitle(pathway)
    
    # Save the plot
    ggsave(filename = file.path(paste0(pathway, "_violin_plot.png")), plot = p, width = 8, height = 6)
    
    print(p)
  
  }
}
```